### PR TITLE
fix: async shutdown handler not getting called

### LINKED ISSUE #964: Shutdown handler not working
### Bug Description

Code in `shutdown_handler` is never executed at shutdown. If a database connection is left open at time of shutdown/restart and `--dev` is given, a process is left hanging, preventing further use of the port.

### Steps to Reproduce

For example, no message is printed upon receiving SIGINT: 

```
from robyn import Robyn

app = Robyn(__file__)

@app.get('/')
async def index(request):
    return 'hello world'

@app.startup_handler
async def startup_handler():
    print('Starting up...')

@app.shutdown_handler
async def shutdown_handler():
    print('Shutting down...')

app.start(port=8081)
```

### Your operating system

Linux

### Your Python version (`python --version`)

3.12

### Your Robyn version

latest

### Additional Info

_No response_

### PR DESCRIPTION
**Pull request by VishnuSanal**
## Description

This PR fixes #964

## Summary

This PR does fix async shutdown handler not getting called

## PR Checklist

Please ensure that:

- [x] The PR contains a descriptive title
- [x] The PR contains a descriptive summary of the changes
- [x] You build and test your changes before submitting a PR.
- [ ] You have added relevant documentation
- [ ] You have added relevant tests. We prefer integration tests wherever possible

## Pre-Commit Instructions:

- [x] Ensure that you have run the [pre-commit hooks](https://github.com/sparckles/robyn#%EF%B8%8F-to-develop-locally) on your PR.

### CODE REVIEW

File: robyn/cli.py [hunk 1]
```diff
@@ -131,4 +131,7 @@ def run():
         print("Starting dev server...")
         start_dev_server(config, config.file_path)
     else:
-        start_app_normally(config)
+        try:
+            start_app_normally(config)
+        except KeyboardInterrupt:
+            pass
```
Comment (sansyrox) [inline robyn/cli.py:137]: why are we doing this?

File: src/server.rs [hunk 2]
```diff
@@ -216,22 +215,17 @@ impl Server {
...
-                pyo3_asyncio::tokio::run(py, async move {
-                    execute_event_handler(shutdown_handler, &task_locals.clone())
-                        .await
-                        .unwrap();
-                    Ok(())
-                })
-            })?;
-            abort();
-        }
+        let _ = (*event_loop).call_method0("run_forever");
```
Comment (sansyrox) [inline src/server.rs:227]: could you still handle any potential errors here?

**Comment by VishnuSanal** [comment]
note to self: Sanskar taking this over after discussion in the meeting. :)

**Comment by VishnuSanal** [comment]
- [ ] fix crash event loop already runnign

File: robyn/cli.py (hunk 1 above)
Comment (VishnuSanal) [reply robyn/cli.py:137]: crashed upon ctrl + c

**Comment by VishnuSanal** [comment]
current issues:

- [ ] have to press ctrl + c twice for an app with async shutdown handler to quit
- [ ] dev server doesn't get terminated properly -- the post is occupied the next time the app runs

**Comment by sansyrox** [comment]
Hey @VishnuSanal ,

There are some conflicts. Can you have a look again?

**Comment by VishnuSanal** [comment]
@sansyrox PTAL

File: robyn/cli.py [hunk 3]
```diff
@@ -129,4 +129,7 @@ def run():
...
```
Comment (sansyrox) [inline robyn/cli.py:135]: @VishnuSanal , could you add a comment here?

File: src/server.rs [hunk 4]
```diff
@@ -222,14 +224,27 @@ impl Server {
...
+                        .unwrap(),
+                    )
+                    .unwrap();
+                } else {
+                    debug!("Shutdown event handler");
+
+                    Python::with_gil(|py| function.handler.call0(py))?;
+                }
+            }
+
```
Comment (sansyrox) [inline src/server.rs:247]: nice

File: src/server.rs [hunk 5]
```diff
@@ -222,14 +224,27 @@ impl Server {
...
+                    pyo3_asyncio::tokio::run_until_complete(
+                        task_locals.event_loop(py),
+                        pyo3_asyncio::into_future_with_locals(
+                            &task_locals.clone(),
+                            function.handler.as_ref(py).call0()?,
+                        )
+                        .unwrap(),
+                    )
+                    .unwrap();
```
Comment (sansyrox) [inline src/server.rs:240]: this should be a separate function. called `execute_shutdown_handler` in the same file as https://github.com/sparckles/Robyn/pull/965/files#diff-e57d319b832ca59f61da275f35a0903e6a4f6a0fbfee4d2cfb149beb1f7aa220

File: robyn/cli.py (hunk 3 above)
Comment (sansyrox) [reply robyn/cli.py:135]: I would like to suggest some comments that explicitly suggest why is this a good solution? Or atleast `crashes on pressing ctrl+c`

File: src/server.rs (hunk 5 above)
Comment (VishnuSanal) [reply src/server.rs:240]: doing this from the same file (i.e, without starting an async function) was the solution to our issue.

now, if I try to create a function & move the code there, I get `unused implementer of `futures_util::Future` that must be used futures do nothing unless you `.await` or poll them`. now, if I add `await`, I get the error ``await` is used inside non-async function, which is not an `async` context`. which can be solved by starting a new async function & we are back to square one.

or, am I missing something? do you have any other suggestions for me to try? 🤔

Comment (sansyrox) [reply src/server.rs:240]: @VishnuSanal , I don't see you awaiting anything in the code. What do you mean by `.await` or polling it here?

Comment (sansyrox) [reply src/server.rs:240]: > doing this from the same file (i.e, without starting an async function) was the solution to our issue.

now, if I try to create a function & move the code there, I get unused implementer of futures_util::Futurethat must be used futures do nothing unless you.await or poll them. now, if I add await, I get the error ``awaitis used inside non-async function, which is not anasync` context`. which can be solved by starting a new async function & we are back to square one.

or, am I missing something? do you have any other suggestions for me to try? 🤔


also mention this in the code

File: src/server.rs [hunk 6]
```diff
@@ -222,14 +224,39 @@ impl Server {
...
+            // to fix an issue that arises when a new async function is spooled up.
+
+            // if I try to create a function & move the code there, I get an error stating
+            // "unused implementer of `futures_util::Future` that must be used futures do nothing
+            // unless you await or poll them."
+
+            // now, if I add `.await`, I get the error "await is used inside non-async function,
+            // which is not an async context".
+
+            // which can be solved by starting a new async function & we are back to square one.
```
Comment (sansyrox) [inline src/server.rs:238]: this needs improvement

**Review by sansyrox — APPROVED** [review]
LGTM! 👍 🚀

**Pull request merged**